source/_posts/2017/2017-09-08-how-to-deprecate-code-functionality -in-a-symfony-way.md

```diff
@@ -0,0 +1,316 @@
+     * @deprecated
+     */
+    public function add(...)
+    {
+        $this->addHtml(...);
+    }
+    
+    public function addHtml(...)
+    {
+        // ...
+    }
+}
+```
+
+This is the least you can do. 
+
+But you could do better, right?
+
+```php
+/**
+ * @deprecated Method add() is deprecated.
+ */
+public function add(...)
+```
+
+Hm. I should probably delete all those methods calls in my code and hope it still works...
+
+...or maybe there is solution to my pain:
+
+```php
+/**
+ * @deprecated Method add() is deprecated, use addHtml() instead.
+ */
+public function add(...)
+```
+
+
+**Nice!** Even I can understand this now and I'm very poor English.
+
+
+
+### I Have 1 Question for you
+
+**What happens when programmer runs `$html->add(...)` method with your code?**
+
+...
+
+Well, exactly... Nothing. It will work and he or she won't even know that there is some deprecation going.
+
+
+*Thanks Captain Obvious.*
+
+
+That's because annotation have no influence on code run.
+
+<br>
+
+Luckily, there is option that **will actually inform about the deprecation**.
+
+
+
+### 2. A `trigger_error()`   
+
+A [`trigger_error()`](http://php.net/manual/en/function.trigger-error.php) is native PHP function, that can inform user about changes in the code. 
+
+With 2nd argument is level of these messages - there is special constant `E_USER_DEPRECATED` destined for this case.
+
+```php
+namespace Nette\Utils;
+
+class Html
+{
+    public function add(...)
+    {
+        # we already know how to write useful mesagges
+        trigger_error('Method add() is deprecated, use addHtml() instead.', E_USER_DEPRECATED);
+        
+        $this->addHtml(...);
+    })
+    
+    public function addHtml(...)
+    {
+        // ...
+    }
+}
+```
+
+
+You can [see it used in similar way](https://github.com/nette/utils/blob/f1584033b5af945b470533b466b81a789d532034/src/Utils/Html.php#L362) in the original code. 
+
+
+### I Have 1 Question for you
+
+**What happens when programmer runs `$html->add(...)` method with your code?** 
+
+2 things:
+
+- The code will run
+- **The programmer will be informed**
+
+
+In case he or she is not ready for upgrade, it can be disabled in application `bootstrap` file:
+
+```php
+error_reporting(~E_USER_DEPRECATED);
+```
+
+[Source](https://phpfashion.com/jak-spravne-updatovat-nette) (Czech only)
+
+
+
+I said...
+
+> It's very simple to add to your open-source code workflow...
+
+...and this is it!
+
+
+
+## Bonus: Gradual Deprecation
+
+You can see [the same technique in Symfony](https://github.com/symfony/symfony/blob/c9bee848c5e009de190a8cc4ca85948a93ded129/src/Symfony/Component/Config/Definition/ArrayNode.php#L254) with 1 difference: 
+
+```php
+@trigger_error('Method add() is deprecated, use addHtml() instead.', E_USER_DEPRECATED);
+```
+
+Notice `@`!
+
+So the error is there for the user, but it is suppressed? WTF!
+
+
+### Step by Step
+
+It's related to release process. Just imagine this life cycle of `add()` method:
```

Comment:
I don't think Symfony uses the at-sign to achieve this. And I don't like such gradual deprecations by the way.

The standard `trigger_error()` call means that the notices are opt-out. You need to disable them with `error_reporting`. Symfony uses at-signs to make them opt-in instead - you will only see them if you use an error handler that would make them visible. See https://symfony.com/doc/current/components/phpunit_bridge.html#trigger-deprecation-notices.

The reason why I don't like this Step by Step approach is that you would then mix both opt-in and opt-out deprecation notices in your code which seems like a very bad practice. Please choose just one and be consistent about it. I would recommend removing this Step by Step section and rewrite it with opt-in vs opt-out behaviour, perhaps with an example of the opt-in error handler.

It's just a metter of opinion which one you prefer. I personally prefer the opt-in approach.

Reply:
> I don't think Symfony uses the at-sign to achieve this. And I don't like such gradual deprecations by the way.

Oh, thanks for the link!

> The reason why I don't like this Step by Step approach is that you would then mix both opt-in and opt-out deprecation notices in your code which seems like a very bad practice.

You're right. It's bit confusing, now when I read Symfony code I see them both with no logic behind that.

>  I would recommend removing this Step by Step section and replace it with opt-in vs opt-out section, perhaps with an example of the opt-in error handler.

I'll probably just remove it. This is more complicated then I thought and personally I use them without `@` everywhere.